File: src/ros2_demo_cpp/src/add_two_ints_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp"

using std::placeholders::_1;
using std::placeholders::_2;

class AddTwoIntsClientNode: public rclcpp::Node
{
public:
    AddTwoIntsClientNode() : Node("add_two_ints_client"){
        RCLCPP_INFO(this->get_logger(), "Add two ints server has been started");

        // thread1_ = std::thread(std::bind(&AddTwoIntsClientNode::callbackAddTwoIntsServer, this, 1, 11));
        threads_.push_back(std::thread(std::bind(&AddTwoIntsClientNode::callbackAddTwoIntsServer, this, 2, 12)));
        threads_.push_back(std::thread(std::bind(&AddTwoIntsClientNode::callbackAddTwoIntsServer, this, 3, 13)));
    }

    void callbackAddTwoIntsServer(int a, int b){
        auto client = this->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");
        while (!client->wait_for_service(std::chrono::seconds(1)))
        {
            RCLCPP_WARN(this->get_logger(), "Waiting for add two ints server");
        }
        auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
        request->a = a;
        request->b = b;
        
        auto future = client->async_send_request(request);

        try{
            auto reponse = future.get();
            RCLCPP_INFO(this->get_logger(), "%d + %d = %d", request->a, b, reponse->sum);
        } catch(std::exception &e){
            RCLCPP_INFO(this->get_logger(), "Service call failed");
        }
    }

    private:
    // std::thread thread1_;
    std::vector<std::thread> threads_;
    rclcpp::TimerBase::SharedPtr timer_;
    std::string robot_name_;
    int counter_;
};


int main(int argc, char **argv){
    rclcpp::init(argc, argv);
    auto node = std::make_shared<AddTwoIntsClientNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/ros2_demo_py/ros2_demo_py/add_two_ints_client.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts
from functools import partial

# create class to excute
class AddTwoIntClientNode(Node):
    def __init__(self):
        super().__init__("add_two_ints_client")
        self.get_logger().info("Hello from Vinh, this is the first test and hello Ros 2")
        self.call_add_two_int_server(111, 222)
        self.call_add_two_int_server(112, 222)
        self.call_add_two_int_server(113, 222)

    def call_add_two_int_server(self, a, b):
        client = self.create_client(AddTwoInts, "add_two_ints")
        while not client.wait_for_service(1.0):
            self.get_logger().warn("waiting for server")

        request = AddTwoInts.Request()
        request.a = a
        request.b = b
        future = client.call_async(request)
        future.add_done_callback(partial(self.add_two_ints_callback, a = a, b = b))

    def add_two_ints_callback(self, future, a, b):
        try:
            reponse = future.result()
            self.get_logger().info(str(a) + " + " + str(b) + 
                                    " = " + str(reponse.sum))
        except Exception as e:
            self.get_logger().error("Service call failed %r" %(e, ))

def main(args = None):
    rclpy.init(args=args)
    node = AddTwoIntClientNode()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == "__main__":
    main()